### ISSUE TITLE
can we exempt `Rocket Request Guards` for DBConnectionPool

### ISSUE THREAD

**Issue by micfan** [SergioBenitez/Rocket#1011]
Hi, I'm a newbie on Rust and Rocket.

I'm so sad for passing the DBConnection argument as a [Rocket Request Guards](https://rocket.rs/v0.4/guide/requests/#request-guards) every where of the Rocket project.

Can we make a DBConnectionPool object as global resource, so that our code can be cleaner?

## Questions

Any questions _must_ include:

  1. The version of Rocket this question is based on, if any.

> rocket = "0.4.0"

  2. What steps you've taken to answer the question yourself.

> stackoverflow

  3. What documentation you believe should include an answer to this question.

> https://rocket.rs/v0.4/guide/state/#databases

## Feature Requests

  1. Why you believe this feature is necessary.
> make code looks better, for a DBConnectionPool, is it necessary to Guard?

  2. A convincing use-case for this feature.
> as I did in other programming language, no Guard needed

  3. Why this feature can't or shouldn't exist outside of Rocket.
> what about: let's save some time?

**Comment by jebrosen** [comment]
> Can we make a DBConnectionPool object as global resource, so that our code can be cleaner?

That's a very subjective call. Many would also say that a global resource is *less* clean, because any code anywhere might depend on it in a hard-to-trace way. This matters especially for situations such as testing or debugging.

> as I did in other programming language, no Guard needed

I'm curious what other programming language(s) or framework(s) you are referring to and how they solve this problem.

> what about: let's save some time?

Globally accessible state might save time when writing code, but it often makes the same code more difficult to read later. It is also not something that Rocket is really in a position to do anyway, without a lot of clever workarounds or hacks.

**Comment by micfan** [comment]
> That's a very subjective call. Many would also say that a global resource is less clean, because any code anywhere might depend on it in a hard-to-trace way. This matters especially for situations such as testing or debugging.

That's a very subjective call +1.

Is this a sample of repeat-of-myself? However, I found the same design in the actix-web framework. 

> I'm curious what other programming language(s) or framework(s) you are referring to and how they solve this problem.

Sorry, I don't know so much. As in Ruby on Rails, Django, or Flask, maybe there is a mutex lock for DBConnPool.

> Globally accessible state might save time when writing code, but it often makes the same code more difficult to read later. It is also not something that Rocket is really in a position to do anyway, without a lot of clever workarounds or hacks.

Is it possible to write a Macro to avoid explicitly declaring the position parameter of shared State, each time creating a route handler. 

This parameters is not Agile or Scrum.

If we want to develop fast, we should not repeat self, let the framework hold everything for us. 

Don't let developer think, like Golang, there is no beautiful features, simple, stupid, but fast and work.

**Comment by SergioBenitez** [comment]
You can absolutely use global variables if you'd like (see [`state`](https://docs.rs/state/0.4.1/state/)), but Rocket absolutely won't help you do so in any way. There are many reasons while global variables are a poor choice, and I won't reiterate most of them as a quick web search is likely to yield many answers, but I will point out one non-obvious reason why using a global variable is a bad idea here.

Consider running two servers in one application, each using a distinct database. Today, this is trivially accomplishable with managed state as follows (in pseudocode):

```rust
#[get("/")]
fn go(db: Db) -> Data { db.execute_query() }

fn main() {
    let app1 = rocket().manage(Db::connect("db1"));
    let app2 = rocket().manage(Db::connect("db2"));
    spawn(|| app1.launch());
    spawn(|| app2.launch()).join();
}
```

The `go` function will execute in any number of threads for either server, always using the appropriate database.

Now, consider a version of the above that uses a global variable, again in pseudocode:

```rust
static DB1 = Db::connect("db1");
static DB2 = Db::connect("db2");

#[get("/")]
fn go() -> Data {
    // Which database do I use?
}

fn main() {
    let app1 = rocket();
    let app2 = rocket();
    spawn(|| app1.launch());
    spawn(|| app2.launch()).join();
}
```

It is impossible to determine which database to use: there is no server-local state indicating which database should be used by a route.

While this exact scenario may seem unlikely, a variant of it occurs when testing: your application is likely to want to use a different database when testing as opposed to executing. It's not clear how using a global variable would allow for this without significant effort or a mechanism similar to managed state.

In short: no, Rocket will not expose APIs to access any state whatsoever globally.
